lib/wise_homex/api_definition.ex

```diff
@@ -0,0 +1,61 @@
+defmodule WiseHomex.ApiDefinition do
+  @moduledoc """
+  The definition of the Wise Home API
+  """
+
+  defmodule APIResource do
+    @moduledoc """
+    A definition of an api resource
+    """
+
+    @type t :: %__MODULE__{
+            name_singular: String.t(),
+            name_plural: String.t(),
+            path: String.t(),
+            type: String.t()
+          }
+
+    defstruct [:name_singular, :name_plural, :path, :type]
+  end
+
+  @type endpoint_name :: atom()
+
+  @type endpoint_definition :: %{endpoints: [atom], path: String.t(), type: String.t()}
+
+  @type api_definition :: %{endpoint_name => endpoint_definition}
+
+  @doc """
+  The definition of the Wise Home API
+  """
+  @spec api_definition() :: api_definition
+  def api_definition() do
+    %{
+      expense: %{
+        endpoints: [:index, :show, :create, :update, :delete],
+        path: "/expenses",
+        type: "expenses"
+      }
+    }
+  end
+
+  @doc """
+  Transforms the api definition into an APIResource struct representing every function that should be created
+  """
+  @spec to_api_resources(api_definition()) :: [APIResource.t()]
+  def to_api_resources(api_definition) do
```

Comment:
Perhaps create a facade for `api_definition() |> to_api_resources()`, since the data (`api_definition`) should not be public if only the resource structs are usable.

Reply:
Good suggestion. Done in https://github.com/wise-home/wise_homex/pull/107/commits/e1deb2774fc7b9d816d0d88d59dba8afa9869b9c